core-rust/state-manager/src/store/rocks_db.rs

```diff
@@ -955,8 +964,28 @@ impl<R: ReadableRocks> ConfigurableDatabase for StateManagerDatabase<R> {
         self.config.enable_local_transaction_execution_index
     }
 
-    fn is_historical_substate_value_storage_enabled(&self) -> bool {
-        self.config.enable_historical_substate_values
+    fn get_first_stored_historical_state_version(&self) -> Option<StateVersion> {
+        if !self.config.enable_historical_substate_values {
+            return None; // state history feature disabled explicitly
+        }
+
+        let first_state_hash_tree_version = self
+            .open_read_context()
+            .cf(StaleStateHashTreePartsCf)
+            .get_first_key();
```

Comment:
`ValuesAssociatedWithStateHashTreeSince` is only recorded when the feature is enabled.

E.g. if enabled at v=1400 and never disabled again, then it simply stays at 1400 forever.
But the JMT's GC keeps working. Let's say the we configured history_length=300, and we are now at v=1900.
The oldest valid version is 1900-300=1600 (all older were GC'ed). And the number 1400 does not even matter anymore (i.e. it stopped being relevant after the GC passed that point).

Or did you mean "why can't we bump the `ValuesAssociatedWithStateHashTreeSince` at the end of GC run?".
We could, it just adds some complexity... it would no longer be "just a static piece of config" persisted on boot.

I can also imagine a solution which checks against `StateHashTreeGcConfig::state_version_history_length` instead. But it conceptually feels a bit wrong:
- it is a config of GC, not of DB (and the DB would have to awkwardly reach for it),
- if someone reconfigures the history length (e.g. changes it from 300 to 700), then this number would be a lie for 400 versions after boot.

Hence, I think that "the beginning of `StaleStateHashTreePartsCf` table" is the source of truth here.
If this becomes a performance problem _[visible_doubt.jpg]_ then we'll have to resort to something like `ExtensionsDataKey::VersionReachedByJmtGc`.

Reply:
Ahh I see I misunderstood this:

> `ValuesAssociatedWithStateHashTreeSince` is only recorded when the feature is enabled.

This feels a little weird to me. I'm not sure why we need to know this value specifically? If I started it at state version 2 but I only have data from state version 500, it's not really relevant that I started it at 2.

In my head, the value we need to know is the current state of the index - which is "from what state version do we have all associated substate values?", which basically starts at the enabled state version, then is bumped up to `latest_stale_disposed_state_version + 1`.

I feel it makes sense saving this at each GC rather than calculating it every time using an implementation detail, but if you feel the opposite, I'm happy to explain here with a comment here and on the field of the `StateHashTreeAssociatedValuesStatus` instead.

Reply:
> the JMT GC can't run when we're doing our first index fixing, right?

exactly - all our backfills run as part of `ActualStateManagerDatabase` constructor (which is not very elegant, but fairly bulletproof way of ensuring that no concurrent process access DB in that time; and it's actually not the first time we rely on this, because I remember some similar discussion with Łukasz, possibly about protocol updates).

> I feel it makes sense saving this at each GC rather than calculating it every time using an implementation detail

huh, looks like the boundary of "impl detail" is quite blurry within the `rocks_db.rs` area :(

Because in my interpretation:

- reading the `StateHashTreeAssociatedValuesStatus.first_available_state_version` (which is maintained by GC via periodic upserts)

is _not different_ from:

- reading the first key of `StaleStateHashTreePartsCf` (which is maintained by GC via periodic deletes)

To resolve this, can I suggest:
Let's merge this as-is, and let's wait for https://github.com/radixdlt/babylon-node/pull/873 to merge too.
And then, with the feature's and GC's tests on board, I will create a probably-actually-trivial refactoring PR which migrates the approach. In there, we will decide if it looks conceptually/practically better before or after the refactor. Is it ok?